### PR TITLE
rework: dispense batching

### PR DESCRIPTION
**Pull request by RafaelTaranto**

### CODE REVIEW

**Comment by naconner** [comment]
[Incorrect logging on dispensed bills when batched](https://trello.com/c/TehUUPAr/104-incorrect-logging-on-dispensed-bills-when-batched)

File: lib/brain.js [hunk 1]
```diff
@@ -3180,7 +3143,81 @@ Brain.prototype._physicalDispense = function _physicalDispense () {
   emit('billDispenserDispensing')
 
   const notesToDispense = optimizeDispense(notes, this.billDispenser.dispenseLimit)
-  dispenseGenerator(notesToDispense, this.tx, this.config.completedTimeout)
+  const batchAmount = notesToDispense.length
+
+  const batchDispense = ({ dispensed, current, error }) => {
```
Comment (joshmh) [inline lib/brain.js]: I think we should break out long functions like this to regular JavaScript functions outside of the function. Better to have byte-sized pieces of logic.

Comment (joshmh) [inline lib/brain.js]: Use more descriptive variable names: ``currentBatch``, ``dispenseRecords``.

File: lib/brain.js [hunk 2]
```diff
@@ -3180,7 +3143,81 @@ Brain.prototype._physicalDispense = function _physicalDispense () {
...
+    // short circuit out if a error is found
+    if (error) {
+      return { dispensed, error }
+    }
+
+    const notes = notesToDispense[current - 1]
```
Comment (joshmh) [inline lib/brain.js]: Rename ``current - 1`` to descriptive variable. ``current - 1`` should be checked to ensure it's in range.

File: lib/brain.js [hunk 3]
```diff
@@ -3180,7 +3143,81 @@ Brain.prototype._physicalDispense = function _physicalDispense () {
...
+
+    this.browser().send({ action: 'dispensing', dispenseBatch: { current, of: batchAmount } })
```
Comment (joshmh) [inline lib/brain.js]: Not for now, but we should consider replacing this kind of thing with an emitted event, like we do with the board manager. Same will be true of events destined for l-s. Of course, even better would be to have a petri net emit this.

File: lib/brain.js [hunk 4]
```diff
@@ -3180,7 +3143,81 @@ Brain.prototype._physicalDispense = function _physicalDispense () {
...
+        this.browser().send({ action: screen, dispenseBatch: { current: current, of: batchAmount } })
+
+        return this.billDispenser.waitForBillsRemoved()
+          .then(() => response)
+          .catch(error => _.assign(response, { error }))
+      })
+      .catch(error => ({ dispensed, error }))
+  }
+
+  const batchesFinished = ({ dispensed, current, error }) => {
```
Comment (joshmh) [inline lib/brain.js]: Break out into own function and describe what it does.

File: lib/brain.js [hunk 5]
```diff
@@ -3180,7 +3143,81 @@ Brain.prototype._physicalDispense = function _physicalDispense () {
...
+      if (tx.id !== _.get('id')(this.tx)) {
+        return
+      }
+      return this._completed(tx.id)
+    })
+  }
+
+  const tasks = [...Array(batchAmount).fill(batchDispense), batchesFinished]
+
+  pWaterfall(tasks, {
```
Comment (joshmh) [inline lib/brain.js]: Describe what this does.

**Pull request merged**